Add connection timeout for WiFiClass::begin(ssid)

mrhavlin opened the pull request:
This change adds a public method WifiClass::begin(char* ssid, uint32_t u32TimeoutMS)

The hard-coded 60 second connection timeout was not working well for my application-- added a timeout argument for station mode connections to open access points.  Also consolidated the 60s timeout number into a private const class member WiFiClass::_connectionTimeout for easier modificaton of the global default connect timeout.  There is currently no bounds checking on the timeout argument.

sandeepmistry commented:
Hi @mrhavlin,

Thanks for taking the time to submit this pull request!

> The hard-coded 60 second connection timeout was not working well for my application

Hmm, what timeout value did you end up using to make things work?

mrhavlin commented:
@sandeepmistry -- I am still testing different values between 5 and 20 seconds to see what the tradeoffs are.  I won't bore you with all the details of my application, but the device is running on a battery and I want to extend the device life as much as I can, so if the connection is going to fail I want it to fail more quickly so the device can go back to a low power state.

sandeepmistry commented:
Hi @mrhavlin,

Please bore us with the details :)

I've done some experiments here with the `ConnectWithWPA` example, modifying `setup()` with:

```arduino
    unsigned long start = millis();
    
    // Connect to WPA/WPA2 network:
    status = WiFi.begin(ssid, pass);

    unsigned long end = millis();

    Serial.print("connection time is: ");
    Serial.println(end - start);
```

Here are some results:

* Successful connection: 1553 ms
* Connection to invalid SSID name: 974 ms (first time), 884 ms (second time)
* Connection with invalid password: 1008 ms (first time), 916 ms (second time)

Based on this, I would say adding a dynamic timeout value would not give us too much improvements for the 5 - 20s values you are trying. The timeout value is more for the DHCP lease response.

mrhavlin commented:
@sandeepmistry -- Thanks for building that test case.  Unfortunately you set up a test case that, predictably, returns very quickly.  I appreciate you trying to help me troubleshoot, but I have already diagnosed my application issue and implemented a solution, which I am sharing.  

In my application I was regularly hitting the 60s timeout in WiFiClass::begin(char*).  This was not acceptable to me as it kept the device in a higher power state and also blocked program execution for 60s.  Whether the delay is in the association, authentication, or waiting for an IP lease to complete after the connection process is really not relevant--the connection fails and wastes quite a bit of time in doing so.  The only remedy was to modify the library code.  I did that, and implemented a solution that worked for me, so I thought I would share it.

sandeepmistry commented:
@mrhavlin fair enough, thanks for sharing.

Let's leave this open and see what the demand is from the community. We're a little cautious about adding another (optional) parameter to `begin` at this time.

ericbaril72 commented:
I did go into the WiFi.begin() to lower the timeout value to 5 seconds.

In my case, the problem is not the timeout per say but the fact that a bad network can freeze your application for so long ( up to 60 seconds ).

I would prefer non-blocking solution ...  something like
loop ()
{
   if ((WIFI.status()!= WL_CONNECTED) && (WIFI.status()!= WL_PENDING_CONNECTION) ){
           Wifi.Asyncbegin();
   }

   else if (Wifi.status()==WL_CONNECTED){
             // client or server code here...
            //
   }
}

I have seen a few wifi routers on the same networks that "share" the same SSID and communicate together to let you connect anywhere in the factory using the same credentials.

However, having lots of connect/disconnect does not always work well ... which results into 60 seconds timeout and application freeze.
Using a different "non-shared" SSID never created this issue with the WINC1500.

WendallH commented:
Interesting... I have a very similar application using the WINC1500 and a factory full of access points all with the same SSID and password... and also have the WINC1500 lock-up. 

In my case the WINC1500 is operating as a server and a client connects to it every 500ms. I have had connection issues and have implemented a timer such that if it doesn't see a client in 5 seconds it can assume the connection is broken. Monitoring using the USB serial port the WiFi.status() indicates it is connected to the AP, though monitoring the AP's log shows it is not connected and it also can not be pinged (by a PC). 

After gathering log files from the AP monitor I found that the WINC1500 was disconnecting (likely because it had not "seen" a client in 5 seconds) and attempts to re-connect were futile.

I ended up adding a wire from the SAMD21 "Feather" I am using to the Reset pin of the WINC1500. According to documentation the reset pin on the WiNC1500 is "controlled by the library" so I use the one I added.

Now when there isn't a client for 5 seconds I use the client.end() and the disconnect() commands and then reset the WiNC1500. TaDa.... they (13 of them) have been running for 4 days without a hitch. More time will tell.

Just thought someone might like to know......

sandeepmistry commented:
Closed via https://github.com/arduino-libraries/WiFi101/pull/265, in the next version of this library there is a new `WiFi.setTimeout(ms)` API to set the connect timeout in milliseconds. This was added by @Rocketct.
